### PR TITLE
Add support for INCLUDE NMODL directive

### PR DESCRIPTION
**Pull request by tristan0x**
This pull-request tackles #300 

* handle the INCLUDE NMODL directive
  * parser now follows the included files and store the result AST
    in the `nmodl::ast::Include` AST node.
* update NmodlVisitor generatede class so that it ignores the `program`
  member of the `nmodl::ast:Include` AST node.
* NmodlDriver:
  * new method `parse_include` in charge of parsing
  * move reporting of parsing error in `NmodlDriver` class.
  * remove useless member variables `lexer` and `parser` to make it completely reentrant.
  * all `parse_*` methods now raise exceptions upon error.
* Improve performance of methods:
  * less lookup in datastructure:
    * `NmodlDriver::get_defined_var_value`
    * `SingletonRandomString::get_random_string`
  * return `const&`: 
    * `SingletonRandomString::get_random_string`

### CODE REVIEW

File: src/parser/nmodl_driver.cpp [hunk 1]
```diff
@@ -24,60 +24,107 @@ std::shared_ptr<ast::Program> NmodlDriver::parse_stream(std::istream& in) {
...
+std::shared_ptr<ast::Include> NmodlDriver::parse_include(const std::string& name,
+                                                         const location& loc) {
+    // Try to find directory containing the file to import
+    const auto directory_path = library.find_file(name);
+
+    // Complete path of file (directory + filename).
+    std::string absolute_path = name;
+
+    if (!directory_path.empty()) {
+        absolute_path = directory_path + std::string(1, utils::pathsep) + name;
```
Comment (iomaganaris) [inline src/parser/nmodl_driver.cpp:69]: Here the `directory_path` I believe should have the path of the `mod` file which included the `mod` file. I need to check also how `mod2c` takes care of this case

Comment (tristan0x) [reply src/parser/nmodl_driver.cpp:84]: I think the issue was FileLibrary : it was missing the directory of the top parsed file in the list of directories to look for .mod files

But now I have another issue:

```
bash-3.2$ bin/nmodl ../test/integration/mod/cabpump.mod
[NMODL] [info] :: Processing ../test/integration/mod/cabpump.mod
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: NMODL Parser Error : can not open file : var_init.inc [Location : 43.9-22]
Abort trap: 6
bash-3.2$ bin/nmodl ../test/integration/mod/cabpump.mod
[NMODL] [info] :: Processing ../test/integration/mod/cabpump.mod
[NMODL] [info] :: Running symtab visitor
[NMODL] [info] :: Running LOCAL to ASSIGNED visitor
[NMODL] [info] :: Running code compatibility checker
[NMODL] [info] :: Running verbatim rename visitor
[NMODL] [info] :: Running KINETIC block visitor
[NMODL] [info] :: Running STEADYSTATE visitor
[NMODL] [info] :: Parsing Units
[NMODL] [info] :: Running cnexp visitor
Assertion failed: (solve_node_symbol != nullptr), function create_solution_expression, file /Users/tristan/src/github.com/BlueBrain/nmodl/src/visitors/solve_block_visitor.cpp, line 45.
Abort trap: 6
```

**Comment by cattabiani** [comment]
I am very puzzled. I was trying to get more log using:

`nmodl /home/katta/projects/nmodl/test/integration/mod/cabpump.mod passes --json-ast

[NMODL] [info] :: Processing /home/katta/projects/nmodl/test/integration/mod/cabpump.mod
[NMODL] [info] :: Running symtab visitor
[NMODL] [info] :: Running LOCAL to ASSIGNED visitor
[NMODL] [info] :: Running code compatibility checker
[NMODL] [info] :: Writing AST into /home/katta/projects/nmodl/test/integration/mod/cabpump.mod
Segmentation fault (core dumped)
`
seg fault. I am checking the debugger but the node looks fine except for some modTokens that are nullptr. I am not sure, may they be nullptr? I am not sure

File: src/parser/nmodl_driver.hpp [hunk 2]
```diff
@@ -75,35 +68,49 @@ class NmodlDriver {
...
     bool verbose = false;
 
     /// root of the ast
     std::shared_ptr<ast::Program> astRoot = nullptr;
 
+    /// The file library for IMPORT directives
+    FileLibrary library{FileLibrary::default_instance()};
+
+    /// The list of open files, and the location of the request.
+    std::map<std::string, location> open_files;
```
Comment (cattabiani) [inline src/parser/nmodl_driver.hpp:81]: nitpick: here we do not need ordering. unordered_map?

File: src/parser/nmodl_driver.cpp [hunk 3]
```diff
@@ -24,60 +24,122 @@ std::shared_ptr<ast::Program> NmodlDriver::parse_stream(std::istream& in) {
...
+    }
+
+    // Detect recursive inclusion.
+    if (open_files.find(absolute_path) != open_files.end()) {
+        std::ostringstream oss;
+        oss << name << ": recursive inclusion.\n"
```
Comment (cattabiani) [inline src/parser/nmodl_driver.cpp:90]: nitpick: use .format here?

File: src/utils/CMakeLists.txt [hunk 4]
```diff
@@ -2,15 +2,17 @@
...
-    ${CMAKE_CURRENT_SOURCE_DIR}/common_utils.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/string_utils.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/common_utils.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/perf_stat.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/perf_stat.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/table_data.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/table_data.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/logger.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/logger.cpp
+    common_utils.cpp
```
Comment (cattabiani) [inline src/utils/CMakeLists.txt:5]: why did you change this?

File: src/utils/file_library.hpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+namespace nmodl {
+
```
Comment (cattabiani) [inline src/utils/file_library.hpp:14]: no doxygen documentation?

File: src/utils/file_library.hpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+class FileLibrary {
+  public:
+    /// An empty library
+    FileLibrary() = default;
+    /**
+     *  Initialize the library with the following path:
```
Comment (cattabiani) [inline src/utils/file_library.hpp:20]: is this comment compliant with hpc coding conventions? where is brief?

File: src/utils/CMakeLists.txt (hunk 4 above)
Comment (tristan0x) [reply src/utils/CMakeLists.txt:5]: I needed to add `common_utils.[ch]pp` and used that opportunity to
* sort the list to improve maintainability
* remove `${CMAKE_CURRENT_SOURCE_DIR}` which is implicit for `add_library` CMake function.

File: src/parser/nmodl_driver.cpp (hunk 3 above)
Comment (tristan0x) [reply src/parser/nmodl_driver.cpp:90]: The error message is constructed with a `nmodl::parser::location` that can work with `std::basic_ostream` but not the _format_ library. I wanted to extend the format library to deal with `nmodl::parser::location` as instructed [here](https://fmt.dev/latest/api.html#formatting-user-defined-types) but realized that every bison parser in NMODL has its owned `location.hpp` file:
* `${CMAKE_BINARY_DIR}/src/parser/diffeq/location.hh`
* `${CMAKE_BINARY_DIR}/src/parser/nmodl/location.hh`
* `${CMAKE_BINARY_DIR}/src/parser/unit/location.hh`

Each of them defines a `location` class, all in the same `nmodl::parser` namespace. There should be either one single `nmodl::parser::location` class used by all parsers or they should all be declared in a dedicated namespace. This is an issue that should be addressed before considering extending the _format_ to deal with `nmodl::parser::location`.

I will create a dedicated issue for this so that we can discuss

File: src/utils/file_library.hpp (hunk 5 above)
Comment (tristan0x) [reply src/utils/file_library.hpp:14]: Fixed in d26e5ef676db7da848eff84ac49ed90efa87538f

File: src/parser/nmodl_driver.hpp (hunk 2 above)
Comment (tristan0x) [reply src/parser/nmodl_driver.hpp:81]: Fixed in d26e5ef676db7da848eff84ac49ed90efa87538f

File: src/parser/nmodl_driver.cpp (hunk 3 above)
Comment (cattabiani) [reply src/parser/nmodl_driver.cpp:90]: thx! Interesting...

**Comment by tristan0x** [comment]
> seg fault. I am checking the debugger but the node looks fine except for some modTokens that are nullptr. I am not sure, may they be nullptr? I am not sure

I can reproduce the issue. the segfault happens in
```
json_printer.cpp:75: back->front().push_back(*current);
```
the `current` local variable is `nullptr`

---
Here is the sequence of things that lead to this issue:
* traversing AST
* ...
* `JSONVisitor::visit_include`
  * ...
  *   `JSONVisitor::visit_program` // new attribute added in this PR
    *  `JSONPrinter::flush()  // special case for `Program`
      * `block = nullptr` 

---
Suggestion: in `visit_program`, only call `flush` is the parent node is `nullptr`

**Comment by tristan0x** [comment]
> > seg fault. I am checking the debugger but the node looks fine except for some modTokens that are nullptr. I am not sure, may they be nullptr? I am not sure
> 
> I can reproduce the issue. the segfault happens in
> 
> ```
> json_printer.cpp:75: back->front().push_back(*current);
> ```
> 
> the `current` local variable is `nullptr`
> 
> Here is the sequence of things that lead to this issue:
> 
> * traversing AST
> * ...
> * `JSONVisitor::visit_include`
>   
>   * ...
>   * `JSONVisitor::visit_program` // new attribute added in this PR
>   * `JSONPrinter::flush()`  // special case for `Program`
>     * `block = nullptr`
> 
> Suggestion: in `visit_program`, only call `flush` if the parent node is `nullptr`


179f5a9 fixes the issue.

**Comment by cattabiani** [comment]
wow, this is the first time we use parent pointers!

File: src/parser/nmodl_driver.hpp (hunk 2 above)
Comment (cattabiani) [reply src/parser/nmodl_driver.hpp:81]: ehm... you changed defined_var but not open_files

**Comment by tristan0x** [comment]
> defined_var became a umap while open_files remained a map (while I asked the change on open_files). Is maybe a lapsus or is there a reason behind?

I changed `defined_var` to unordered map before your comment. You were right for `open_files`, key ordering is not required.

**Comment by tristan0x** [comment]
> > defined_var became a umap while open_files remained a map (while I asked the change on open_files). Is maybe a lapsus or is there a reason behind?
> 
> I changed `defined_var` to unordered map before your comment. You were right for `open_files`, key ordering is not required.

Never mind, `defined_var` was changed *after* your comment @cattabiani :)

**Comment by pramodk** [comment]
Hello @tristan0x , I will review this tomorrow but one thing that I would like to change here is to have only one `Program` node for whole AST i.e. if I do:

```
./bin/nmodl ../test/integration/mod/ispc_rename.mod passes --nmodl-ast --json-ast
```

then we get:

```
# tmp/ispc_rename.ast.json

{
  "Program": [
    {
      "NeuronBlock": [
        {
          "StatementBlock": [
....
    {
      "Include": [
        {
          "String": [
            {
              "name": "\"var_init.inc\""
            }
          ]
        },
        {
          "Program": [
            {
              "BlockComment": [
                {
                  "String": [
                    {
                      "name": "\n    Example file to test INCLUDE keyword\n    in NMODL. This file is included in\n    cabpump.mod and takes care of   initializing\n    the dummy variable named \"var\"\n"
                    }
                  ]
                }
              ]
            },
            {
              "FunctionBlock": [
```

we have two Program nodes in whole AST. For internal AST of included node, I suggest to either replace Program node with something else (e.g. Block) or just attach all children inside Include. The reason for this is : At conceptual level, the included file is not the whole "Program" but typically just helper functions. I would like to keep a notion of "Program" representing whole resultant mod file and hence single Program node in whole AST.

**Comment by cattabiani** [comment]
I totally subscribe @pramodk view. I think that the problem highlighted by the integration tests and ioannis is due to having a program node that is not root in the ast. symtab visitor is confused (probably). It would be much easier (even conceptually) to just replace program with its blocks.

**Comment by tristan0x** [comment]
> I totally subscribe @pramodk view. I think that the problem highlighted by the integration tests and ioannis is due to having a program node that is not root in the ast. symtab visitor is confused (probably). It would be much easier (even conceptually) to just replace program with its blocks.

I applied your suggestion @pramodk and @cattabiani

**Comment by tristan0x** [comment]
New error tests to add :
* [ ] filename is empty
   ```
   $ echo '#include ""' | | g++ -x c++ -
   <stdin>:1:10: error: empty filename
   #include ""
            ^
   1 error generated.
   ```
* [ ] filename does not exist
* [ ] included filename has parsing issues

**Comment by pramodk** [comment]
reset this please

**Comment by pramodk** [comment]
retest this please

**Review by cattabiani — APPROVED** [review]

**Comment by cattabiani** [comment]
please retest

**Comment by tristan0x** [comment]
retest this please Jenkins

**Comment by cattabiani** [comment]
please retest

**Comment by tristan0x** [comment]
retest this please Jenkins

File: src/language/templates/visitors/nmodl_visitor.cpp [hunk 7]
```diff
@@ -52,7 +52,9 @@ using namespace ast;
 
 
 {%- macro add_child(node, child) -%}
-    {% if child.nmodl_name %}
+    {% if node.class_name == node_info.INCLUDE_NODE and child.varname == "blocks" %}
```
Comment (pramodk) [inline src/language/templates/visitors/nmodl_visitor.cpp:55]: Just curious what happens with the block : does extra block scope get printed ? i.e.

```
{
    INCLUDE ...
}
```
?

File: test/integration/mod/cabpump.mod [hunk 8]
```diff
@@ -4,7 +4,8 @@ NEURON {
         SUFFIX cadyn
         USEION ca READ cai,ica WRITE cai 
         RANGE ca 
-	GLOBAL depth,cainf,taur 
+	GLOBAL depth,cainf,taur
+		RANGE var
```
Comment (pramodk) [inline test/integration/mod/cabpump.mod:9]: `nitpick` : can we fix the indent?

File: test/unit/parser/parser.cpp [hunk 9]
```diff
@@ -142,6 +144,28 @@ SCENARIO("NMODL parser running number of invalid NMODL constructs") {
...
+                            Catch::Contains("can not open file : unknown.file"));
+    }
+
+    GIVEN("An invalid included file") {
+        nmodl::utils::TempFile included("included.file",
+                                        nmodl_invalid_constructs.at("title_1").input);
+        REQUIRE_THROWS_WITH(is_valid_construct("INCLUDE \"included.file\""),
+                            Catch::Contains("unexpected End of file"));
+    }
+}
```
Comment (pramodk) [inline test/unit/parser/parser.cpp:167]: nice!

File: src/parser/nmodl_driver.hpp [hunk 10]
```diff
@@ -67,43 +60,58 @@ class NmodlParser;
...
     bool trace_scanner = false;
 
     /// enable debug output in the bison parser
     bool trace_parser = false;
 
-    /// pointer to the lexer instance being used
-    NmodlLexer* lexer = nullptr;
-
-    /// pointer to the parser instance being used
-    NmodlParser* parser = nullptr;
```
Comment (pramodk) [inline src/parser/nmodl_driver.hpp:82]: I no longer remember why these members were added. May be during initial testing when I wanted to access lexer/parser instance from driver? But if it's not used then good to remove ! :+1:

File: src/parser/nmodl_driver.cpp [hunk 11]
```diff
@@ -24,60 +24,132 @@ std::shared_ptr<ast::Program> NmodlDriver::parse_stream(std::istream& in) {
...
+        if (loc != nullptr) {
+            parse_error(*loc, oss.str());
+        } else {
+            throw std::runtime_error(oss.str());
+        }
+    }
+
+    auto current_stream_name = stream_name;
+    stream_name = filename;
+    auto absolute_path = utils::cwd() + utils::pathsep + filename;
```
Comment (pramodk) [inline src/parser/nmodl_driver.cpp:51]: A quick question - `absolute_path` here is the dir where nmodl is launched or the directory of input mod file? In case of MOD file usage in scientific projects, IIRC,  `INCLUDE` file paths are related to the `.mod`  file that is including the file. For example, if `/a/b/c/x.mod` includes `INCLUDE "y.inc"` then `y.inc` is in `/a/b/c` directory.

Can you check if it's the same assumption here?

**Review by pramodk — COMMENTED** [review]
Apart from few minor comments/questions, this is ready to merge!

**Comment by cattabiani** [comment]
This branch passed and is ready to be merged. Do you think we need to recheck it? @tristan0x

File: src/language/templates/visitors/nmodl_visitor.cpp (hunk 7 above)
Comment (tristan0x) [reply src/language/templates/visitors/nmodl_visitor.cpp:55]: No, this change is to prevent including the result of the parsed file and just print the filename. The print visitor won't add a block.

File: src/parser/nmodl_driver.cpp (hunk 11 above)
Comment (tristan0x) [reply src/parser/nmodl_driver.cpp:51]: The specification of the INCLUDE directive is different from what you say
https://neuronsimulator.github.io/nrn/py_doc/modelspec/programmatic/mechanisms/nmodl.html#include

> If the full path to the file is not given, the file is first looked for in the current working directory, then in the directory where the original .mod file was located, and then ...

And this behavior is implemented in `NmodlDriver::parse_include` member method.

**Comment by tristan0x** [comment]
> This branch passed and is ready to be merged. Do you think we need to recheck it? @tristan0x

It is ok to be merged on my side. @pramodk had additional questions I just answered, so it is up to him.

File: src/parser/nmodl_driver.cpp (hunk 11 above)
Comment (pramodk) [reply src/parser/nmodl_driver.cpp:51]: 👍

File: test/integration/mod/cabpump.mod (hunk 8 above)
Comment (pramodk) [reply test/integration/mod/cabpump.mod:9]: I see that there are more issues with the indentation in this mod file. Will re-indent whole file separately.

**Review by pramodk — APPROVED** [review]
